tests/testapp/tests/test_controller_methods.py

```diff
@@ -0,0 +1,254 @@
+        (self.current_id, _) = InstanceIDModel.get_or_create_current_instance()
+        self.mc = MorangoProfileController('facilitydata')
+        self.fac1 = FacilityModelFactory(name='school')
+        self.mc._serialize_into_store()
+        self.old_rmc = RecordMaxCounter.objects.first()
+
+    def test_new_rmc_for_existing_model(self):
+        with mock.patch('platform.platform', return_value='Windows'):
+            (new_id, _) = InstanceIDModel.get_or_create_current_instance()
+
+        Facility.objects.update(name='facility')
+        self.mc._serialize_into_store()
+        new_rmc = RecordMaxCounter.objects.get(instance_id=new_id.id, store_model_id=self.fac1.id)
+        new_store_record = Store.objects.get(id=self.fac1.id)
+
+        self.assertEqual(new_rmc.counter, new_store_record.last_saved_counter)
+        self.assertEqual(new_rmc.instance_id, new_store_record.last_saved_instance)
+
+    def test_update_rmc_for_existing_model(self):
+        Facility.objects.update(name='facility')
+        self.mc._serialize_into_store()
+
+        # there should only be 1 RecordMaxCounter for a specific instance_id and a specific model (unique_together)
+        self.assertEqual(RecordMaxCounter.objects.filter(instance_id=self.current_id.id, store_model_id=self.fac1.id).count(), 1)
+
+        new_rmc = RecordMaxCounter.objects.get(instance_id=self.current_id.id, store_model_id=self.fac1.id)
+        new_store_record = Store.objects.get(id=self.fac1.id)
+
+        self.assertEqual(self.old_rmc.counter + 1, new_rmc.counter)
+        self.assertEqual(new_rmc.counter, new_store_record.last_saved_counter)
+        self.assertEqual(new_rmc.instance_id, new_store_record.last_saved_instance)
+
+    def test_new_rmc_for_non_existent_model(self):
+        with mock.patch('platform.platform', return_value='Windows'):
+            (new_id, _) = InstanceIDModel.get_or_create_current_instance()
+
+        new_fac = FacilityModelFactory(name='college')
+        self.mc._serialize_into_store()
+        new_rmc = RecordMaxCounter.objects.get(instance_id=new_id.id, store_model_id=new_fac.id)
+        new_store_record = Store.objects.get(id=new_fac.id)
+
+        self.assertNotEqual(new_id.id, self.current_id.id)
+        self.assertEqual(new_store_record.last_saved_instance, new_rmc.instance_id)
+        self.assertEqual(new_store_record.last_saved_counter, new_rmc.counter)
+
+
+class DeserializationFromStoreIntoAppTestCase(TestCase):
+
+    def setUp(self):
+        DatabaseIDModel.objects.create()
+        (self.current_id, _) = InstanceIDModel.get_or_create_current_instance()
+        self.range = 10
+        self.mc = MorangoProfileController('facilitydata')
+        for i in range(self.range):
+            self.ident = uuid.uuid4().hex
+            StoreModelFacilityFactory(pk=self.ident, serialized=serialized_facility_factory(self.ident))
+
+    def test_dirty_store_records_are_deserialized(self):
+        self.assertFalse(Facility.objects.all())
```

Comment:
Here and elsewhere, when doing this `assertFalse` type of thing on querysets, might be clearer, and marginally faster, for the queryset to use `.exists()`.